CMakeLists.txt

```diff
@@ -1,10 +1,16 @@
-cmake_minimum_required(VERSION 3.16 FATAL_ERROR)
+cmake_minimum_required(VERSION 3.22 FATAL_ERROR)
 
 # scan source files for importing C++20 modules
 if(Kokkos_ENABLE_EXPERIMENTAL_CXX20_MODULES)
   cmake_policy(SET CMP0155 NEW)
 endif()
 
+# Work around MSVC+CUDA issue
+cmake_policy(SET CMP0126 OLD)
+
+# Work around LANGUAGE setting on source files
+cmake_policy(SET CMP0119 OLD)
```

Comment:
```diff
--- a/cmake/kokkos_arch.cmake
+++ b/cmake/kokkos_arch.cmake
@@ -243,12 +243,12 @@ mark_as_advanced(Kokkos_IMPL_AMDGPU_LINK)
 #clear anything that might be in the cache
 global_set(KOKKOS_AMDGPU_OPTIONS)
 if(KOKKOS_ENABLE_HIP)
+  global_append(KOKKOS_AMDGPU_OPTIONS -xhip)
   set(AMDGPU_ARCH_FLAG "--offload-arch")
   if(NOT KOKKOS_CXX_COMPILER_ID STREQUAL HIPCC)
     if(NOT CMAKE_CXX_STANDARD)
       message(FATAL_ERROR "Kokkos requires CMAKE_CXX_STANDARD to set to 20 or higher")
     endif()
-    global_append(KOKKOS_AMDGPU_OPTIONS -xhip)
     if(DEFINED ENV{ROCM_PATH})
       global_append(KOKKOS_AMDGPU_OPTIONS --rocm-path=$ENV{ROCM_PATH})
     endif()
```
(and similarly for Cuda) might be a more robust solution overriding `-x c++` which causes the issue.